Lib: Util: String conversion API

Andrewpini opened the pull request:
Adds string conversion API.
Implements basic string to parameter convertion functions.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

PavelVPV commented on include/bluetooth/mesh/shell.h:
```diff
@@ -35,6 +37,69 @@ extern struct bt_mesh_health_cli bt_mesh_shell_health_cli;
+ *
+ * @return Converted uint8 value
+ */
+uint8_t mesh_shl_str2u8(const char *str);
```
How about `bt_mesh_shell_*` to align with the other names?

PavelVPV commented on subsys/bluetooth/mesh/shell_utils.c:
```diff
@@ -0,0 +1,155 @@
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <shell/shell.h>
```
Missing `#include "shell_utils.h"`

Andrewpini replied:
I'm not quite sure if I understand. You mean `#include <bluetooth/mesh/shell.h>`?

PavelVPV replied:
Yes, sorry, `bluetooth/mesh/shell.h`.

LingaoM commented on subsys/bluetooth/mesh/shell_utils.c:
```diff
@@ -0,0 +1,155 @@
+	}
+
+	for (int i = idx; i < str_len; i++) {
+		if (!isdigit((int)str[i])) {
+			shell_error(shl,
+				    "\"%s\" is not recognized as a valid signed decimal input",
+				    str);
+			return -EINVAL;
+		}
```
Suggest:

```suggestion
		if (isdigit((int)str[i])) {
			continue;
		}

		shell_error(shl, "\"%s\" is not recognized as a valid signed decimal input", str);

	        return -EINVAL;
```

Andrewpini commented:
After talking to @PavelVPV we have agreed that this API should not be public. Further I have decided that the changes to subsys\bluetooth\shell.c should be done later as a separate task since there are additional changes that needs to be done.

alwa-nordic reviewed (dismissed):
I would like these API additions to go into the shell subsystem. They are not mesh-specific and I think others could benefit. And it will hopefully encourage a uniform interface.

PavelVPV commented on subsys/shell/Kconfig:
```diff
@@ -237,6 +237,12 @@ config SHELL_LOG_BACKEND
 	  using the shell backend's LOG_LEVEL option
 	  (e.g. CONFIG_SHELL_TELNET_INIT_LOG_LEVEL_NONE=y).
 
+config SHELL_STRING_CONV
```
Not sure if a separate Kconfig option is required in this case. The functions will anyway be removed from the code if they are not used.

PavelVPV commented on subsys/shell/shell_string_conv.c:
```diff
@@ -0,0 +1,159 @@
+ */
+
+#include <shell/shell.h>
+#include <shell/shell_string_conv.h>
```
Shouldn't this include be moved inside `shell.h`?

alxelax reviewed (commented):
LGTM, agree with @PavelVPV comments

jakub-uC commented on subsys/shell/shell_string_conv.c:
```diff
@@ -0,0 +1,159 @@
+
+int shell_str2bool(const struct shell *shl, char *str, bool *val)
+{
+	if (!strcmp(str, "on") || !strcmp(str, "enable")) {
```
May you please let me know what can be a use case for this function?
I ask out of curiosity because usually options like `on` and `enable` are implemented as subcommands.

Andrewpini replied:
It is just something that is inherited from the mesh shell module. These functions will eventually replace the string conversions inside the mesh shell module. On second thought it would be cleaner to leave this out and create a wrapper for it where it is needed.

nordic-krch commented on include/shell/shell_string_conv.h:
```diff
@@ -0,0 +1,85 @@
+uint8_t shell_hexstr2num(const struct shell *shl, char *str, uint8_t *buf, uint8_t buf_len);
+
+/**
+ * @brief Convert string to boolean param.
```
work mentioning that `on` and `enable` are also converted to `true`.

nordic-krch commented on subsys/shell/shell_string_conv.c:
```diff
@@ -0,0 +1,159 @@
+
+int shell_str2bool(const struct shell *shl, char *str, bool *val)
+{
+	if (!strcmp(str, "on") || !strcmp(str, "enable")) {
```
if positive options are handled that way, what about `off` and `disable`.

nordic-krch reviewed (commented):
imo, it shouldn't be in shell since it is generic utility for strings. Shell is only used for printing errors. You can use logging instead and decouple if from shell. I would put it somewhere in `lib` maybe `lib/util`. There is already `fnmatch` which is also about patterns in strings.

Please add some tests.

nordic-krch commented:
Looks good but tests are needed as now code is not compiled by CI.

Andrewpini commented:
> Looks good but tests are needed as now code is not compiled by CI.

Im on it 👍

alwa-nordic commented on lib/util/string_conv/string_conv.c:
```diff
@@ -0,0 +1,156 @@
+			continue;
+		}
+
+		LOG_ERR("\"%s\" is not recognized as a valid signed decimal input", str);
```
I'm not sure we should log an error here. I think it would be better is this was a pure function.

Andrewpini replied:
You mean log at all, or in this particular function?

alwa-nordic reviewed (changes requested):
I have some comments. I also think we need unit tests for these functions.

de-nordic commented on lib/util/string_conv/string_conv.c:
```diff
@@ -0,0 +1,156 @@
+		return -EINVAL;
+	}
+
+	temp_val = strtol(str, NULL, 10);
+
+	if ((temp_val <= LONG_MIN) || (temp_val >= LONG_MAX)) {
```
`temp_val` is `int` `strtol` returns `long`, there is a chance that the if does not work as expected because you will already have value truncated here by implicit cast that happens on assignment to `temp_val`.

de-nordic commented on lib/util/string_conv/string_conv.c:
```diff
@@ -0,0 +1,156 @@
+
+int string_conv_str2long(char *str, int *val)
+{
+	int str_len = strlen(str);
+	int temp_val;
+	int idx = 0;
+
+	if (!strncmp(str, "-", 1)) {
+		idx = 1;
+	}
+
+	for (int i = idx; i < str_len; i++) {
+		if (isdigit((int)str[i])) {
```
`strlen` in line 17 is not needed as the end of string may be found in loop 25, this way preventing looping twice the same range of `str`.
`strncmp` for single character seems overkill.

de-nordic commented on lib/util/string_conv/Kconfig:
```diff
@@ -0,0 +1,15 @@
+menuconfig STRING_CONV
+	bool "Enable utility string conversion tools"
+	help
+	  Library for conversion of strings into standard c types.
```
C is proper name and should be capitalized.

de-nordic commented on lib/util/string_conv/string_conv.h:
```diff
@@ -0,0 +1,79 @@
+ *
+ * @return 0 on success, negative error code on failure
+ */
+int string_conv_str2long(char *str, int *val);
```
What is proper string for conversion? Can it start and end with white characters?
Is `   -10   ` proper number?

Andrewpini replied:
In its current state the library does not trim heading or trailing whitespace. Strings with whitespace will be rejected.

de-nordic commented on lib/util/string_conv/string_conv.h:
```diff
@@ -0,0 +1,79 @@
+ *
+ * @return 0 on success, negative error code on failure
+ */
+int string_conv_str2dbl(const char *str, double *val);
```
What is proper string for conversion? Is `  -10.20    ` acceptable?

Andrewpini replied:
"-10.20" is fine, "-10.20 " is not.

de-nordic replied:
So the user would have to trim the string to the acceptable value anyway?
So basically a user would have to recognize whether a string is acceptable then the function will check whether string is acceptable, and then it will make dbl of it.
Wouldn't it be better to just allow it to skip white chars that immediately precede or follow the number and return, similarly to `strtol` pointer to last unprocessed character and let user decide whether it is acceptable that not all characters in buffer were consumable?

Andrewpini replied:
You have a point 👍 I will add func to ignore heading/trailing whitespace.

de-nordic commented on lib/util/string_conv/string_conv.c:
```diff
@@ -0,0 +1,156 @@
+
+	frac_dbl = (double)frac;
+
+	for (int i = 0; i < strlen(frac_buf); i++) {
```
This, in bad scenario, evaluates `strlen` every loop run. Compiler may never be sure whether `frac_buf` is modified in background.

de-nordic replied:
End there is a lot of parsing and counting the same things allover again; for example `frac_buf` is filled by string_conv_str2ulong` that basically checks how many digits are there before then running `str2ulong` that does the same thing again and then we get here where the `frac_buf` gets processed to count digits.

de-nordic commented on lib/util/string_conv/string_conv.c:
```diff
@@ -0,0 +1,156 @@
+	}
+
+	strncpy(deci_buf, str, idx);
+	strncpy(frac_buf, (str + (idx + 1)), MIN(sizeof(frac_buf) - 1, strlen(str + (idx + 1))));
```
`strlen` is redundant as `strncpy` will stop at 0 in str anyway, unless the `N` in `strNcpy` is met first.

Andrewpini commented on lib/util/string_conv/string_conv.h:
```diff
@@ -0,0 +1,79 @@
+ *
+ * @return 0 on success, negative error code on failure
+ */
+int string_conv_str2bool(char *str, bool *val);
```
Should we just toss it? After removing the previous "enabled" and "on" additions it does not really provide anything that can't be achived by one of the other conversion functions and a typecast.

Andrewpini commented on lib/util/string_conv/string_conv.h:
```diff
@@ -0,0 +1,79 @@
+ *
+ * @return Converted buffer size on success, 0 on failure
+ */
+uint8_t string_conv_hexstr2num(char *str, uint8_t *buf, uint8_t buf_len);
```
Well one first look it seems that you are right. I will test it, and if your assumption is correct I will remove it.

PavelVPV commented on lib/util/string_conv/string_conv.c:
```diff
@@ -0,0 +1,172 @@
+	memcpy(out, str, out_size - 1);
+	out[out_size - 1] = 0;
+
+	return out_size;
```
`out_size` is `size_t` while return type is `int`.

PavelVPV commented on tests/lib/string_conv/src/main.c:
```diff
@@ -0,0 +1,223 @@
+	long test_val = 0;
+
+	const struct obj_array ctx = {
+		.elems = {
```
Why not just do `struct input elems[] = { ... };` and then user `ARRAY_SIZE(elems)` instead of `elem_cnt`?

nordic-krch commented on lib/util/string_conv/string_conv.h:
```diff
@@ -0,0 +1,56 @@
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
please move it somewhere to `zephyr/include`, maybe `include/sys`. I know that fnmatch has header in its own directory but it is rather an exception and i don't know what is the reason for that exception.

Andrewpini replied:
Ok, I will do that

nordic-krch commented on lib/util/string_conv/string_conv.h:
```diff
@@ -0,0 +1,56 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef STRING_CONV_H__
```
use name that contains path as well. for example https://github.com/zephyrproject-rtos/zephyr/blob/c31cb0075ad1933da5487db734c118cd5083ae3d/include/sys/cbprintf.h#L7

nordic-krch commented on lib/util/string_conv/string_conv.h:
```diff
@@ -0,0 +1,56 @@
+ * @param str Input string
+ * @param val Converted value
+ *
+ * @return 0 on success, negative error code on failure
```
If possible, be specific about error code.

nordic-krch commented on tests/lib/string_conv/testcase.yaml:
```diff
@@ -0,0 +1,6 @@
+tests:
+  libraries.string_conv:
+    tags: string to numeric conversion
+    platform_allow: native_posix
```
this can be removed. We expect that it works on all platforms.

Andrewpini replied:
Well, I expected that aswell but there are approximatley 10 platforms that fails the unit test. Most of these are related to QEMU platform. My thought was that as long as it runs on the native linux platform, it should cover most other platforms.

nordic-krch commented on lib/util/string_conv/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+#
+
+zephyr_include_directories_ifdef(
+  CONFIG_STRING_CONV
```
can be removed when header is moved.

nordic-krch commented on tests/lib/string_conv/testcase.yaml:
```diff
@@ -0,0 +1,6 @@
+tests:
+  libraries.string_conv:
+    tags: string to numeric conversion
```
Some qemu's has no FPU. I would recommend following: 
1. Turn this test case to one with FPU (remove `CONFIG_FPU=y` from `prj.conf`).
```suggestion
  tags: string to numeric conversion
  filter: CPU_HAS_FPU
  extra_configs:
    - CONFIG_FPU=y
```

2. Add second test case without fpu. This should not need any filter.

In `main.c` add ifdef to fpu cases.

Andrewpini replied:
It was not only the FPU option that was the issue. I tried to filter the test to only run on devices that contained FPU, but that did not resolve this. All three tests for str2long, str2ulong and str2dbl failed. on basis of the feedback from these tests I suspect that it could be an errno related issue for the qemu platforms, but im not certain.

de-nordic commented on lib/util/string_conv/string_conv.c:
```diff
@@ -0,0 +1,176 @@
+	}
+
+	/* Avoid doing str2long if decimal part is empty */
+	if (!strlen(trimmed_buf)) {
```
Just check if `trimmed_buf[0] == '\0'`; there is entire function call involved to only check that, seems to be overkill.

de-nordic replied:
Everywhere where `strlen` call result is directly cast to boolean, it can probably be replaced with direct null terminator check.

de-nordic commented on lib/util/string_conv/string_conv.c:
```diff
@@ -0,0 +1,176 @@
+	long temp_val;
+	int idx = 0;
+
+	if (!strncmp(trimmed_buf, "-", 1) || !strncmp(trimmed_buf, "+", 1)) {
```
Calling `strncmp` where `n` is literal 1 is waste of call. This basically builds up entire function call to just do `return !(trimmed_buf[0] == '-')`

Andrewpini commented:
Since this PR has had major alterations along the way I would like to summarize its current state. As of now I have put the header file in the include/sys folder of zephyr and removed all logging in the module. After removing log I did not see a reason for keeping the Kconfig for this module, so it has been removed, as well as the conditional include for the module in cmake. 

When it comes to the unit test it is currently only running on the native posix platform. Initially the test was run on all available platforms, but a few (10ish) platforms could not pass the unit test. The majority of these failed tests occured on the QEMU platform, where it seems like there is some issue related to the use of errno inside the string conversion module. At this point I can't see any obvious error in the implementation that would cause the test to fail on these few platforms, and I can not see any resonable fix to resolve it. So my question is if it is acceptable to only run the unit test on native posix, since I belive it will cover most cases?

Andrewpini commented:
Link to test run where test was run on all platforms with a FPU. 
Notice that the test fails to return an error on the illegal boundary values that is covered by errno in the implementation :

https://github.com/zephyrproject-rtos/zephyr/runs/5446252959?check_suite_focus=true

nordic-krch commented:
@Andrewpini note that tests are failing on 64 bit platforms. You probably need to handle that in the test.

nordic-krch commented:
e.g. "-2147483649" > "-2147483649L"

Andrewpini committed:
Lib: Util: String conversion API

Adds string conversion API.
Implements basic string to parameter convertion functions.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

PavelVPV commented on include/sys/string_conv.h:
```diff
@@ -0,0 +1,62 @@
+ * @return -EINVAL on invalid string input.
+ * @return -ERANGE if numeric string input is to large to convert.
+ */
+int string_conv_str2long(char *str, long *val);
```
```suggestion
int string_conv_str2long(const char *str, long *val);
```

PavelVPV replied:
Same for the function definition

PavelVPV commented on include/sys/string_conv.h:
```diff
@@ -0,0 +1,62 @@
+ * @return -EINVAL on invalid string input.
+ * @return -ERANGE if numeric string input is to large to convert.
+ */
+int string_conv_str2ulong(char *str, unsigned long *val);
```
```suggestion
int string_conv_str2ulong(const char *str, unsigned long *val);
```

PavelVPV replied:
Same for the function definition

Andrewpini commented:
Any CO that wants to add something? If there is no more dire requests for changes I would appreciate approval of this since it is a blocker for other tasks.

carlescufi commented:
Is there a particular reason why you are not augmenting Zephyr's minimal C library instead?

carlescufi commented:
@de-nordic and @alwa-nordic please take another look

stephanosio reviewed (changes requested):
I am not convinced why we need non-standard functions like this when there already exist the C standard functions to do this:
```
string_conv_str2long - strtol
string_conv_str2ulong - strtoul
string_conv_str2dbl - strtod
```

`strtol` and `strtoul` are already part of the minimal libc.
`strtod` is currently not part of the minimal libc (it is available if you use newlib though).

de-nordic commented:
> Is there a particular reason why you are not augmenting Zephyr's minimal C library instead? i.e. adding `sscanf()` to `lib/libc/minimal`?

Maybe that should be done instead.

Andrewpini commented:
> I am not convinced why we need non-standard functions like this when there already exist the C standard functions to do this:
> 
> ```
> string_conv_str2long - strtol
> string_conv_str2ulong - strtoul
> string_conv_str2dbl - strtod
> ```
> 
> `strtol` and `strtoul` are already part of the minimal libc. `strtod` is currently not part of the minimal libc (it is available if you use newlib though).

Well for my own part I do not have a strong opinion on this, other than that what I am trying to accomplish is not a unique usecase. This is the original description of the task that caused the PR: 
```
The way the Bluetooth mesh shell clients extracts numeric values is vulnerable to garbage input strings.
In its current state a invalid numeric string will in most cases be translated to the value 0, 
which may corrupt the state of a device unintentionally. To improve this the parsing of 
numeric strings should be guarded by checks that ensure that the input is a valid 
numeric string, and prevent further command execution if it is not a valid numeric string.
```
So the main objective I want to acheive is to get error feedback on conversion of corrupt numeric strings. The implementation of strtol has, to the best of my knowledge, only error feedback on conversion overflow. 

Initially this was placed as utility under the shell module for Bluetooth Mesh. Then it was pointed out that this was not mesh spesific functionality, and that it should be general shell utility. Then it was pointed out that this was not a shell spesific usecase either, and here we are. I guess moving it back to where I started is one viable option.

stephanosio commented:
> The implementation of strtol has, to the best of my knowledge, only error feedback on conversion overflow.

`strto*` supports error reporting through `errno` (`ERANGE` and `EINVAL`) and `endptr` (e.g. it will be `NULL` if no conversion is performed).

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/strtol.html

Although the minimal libc implementation may not be fully compliant to the specification linked above at this time, it is important to note that the `strto*` functions themselves are capable of providing input error checks.

alwa-nordic commented:
> Initially this was placed as utility under the shell module for Bluetooth Mesh. Then it was pointed out that this was not mesh spesific functionality, and that it should be general shell utility. Then it was pointed out that this was not a shell spesific usecase either, and here we are. I guess moving it back to where I started is one viable option.

Please, no. If these functions really are safer to use, they should be available and preferred everywhere.

Thalley commented:
> > Initially this was placed as utility under the shell module for Bluetooth Mesh. Then it was pointed out that this was not mesh spesific functionality, and that it should be general shell utility. Then it was pointed out that this was not a shell spesific usecase either, and here we are. I guess moving it back to where I started is one viable option.
> 
> Please, no. If these functions really are safer to use, they should be available and preferred everywhere.

We could add them as wrappers around the existing functions, instead of re-inventing the logic. Basically just use the existing functions and add additional error handling.

alwa-nordic commented:
I believe there is merit in a simpler API. `strtol`, `strtoul` and `strtod` are quite general, which is nice when we want to parse something complicated. But it can be complicated to use. E.g: Matching a decimal number with nothing around is surprisingly tedious. The following code demonstrates what is required.

```c
/* The value of `out` is undefined in case of error. */
int parse_dec_to_long(long * out, char * str) {
        if (isspace(str[0])) {
                return -EINVAL;
        }
        char * endptr;
        errno = 0;
        *out = strtol(str, &endptr, 10);
        if (endptr[0] != '\0') {
                return -EINVAL;
        }
        return -errno;
}
```

Ignoring white space is not always the right thing to do. E.g. for protocols. But even when the user of `strtol` wants to ignore white space, they still need to check for garbage in `endptr`. (I.e. `strtol("123abc", NULL, 10)` will evaluate to `123` and not set an errno value.)

We should also consider the parsing of non-null-terminated strings. This cannot be easily done with `strtol`.
